ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.hpp

```diff
@@ -0,0 +1,229 @@
+    size_t receiver_num_buffers;
+
+    size_t local_sender_channel_0_buffer_address;
+    size_t local_sender_channel_0_connection_info_addr;
+    size_t local_sender_channel_1_buffer_address;
+    size_t local_sender_channel_1_connection_info_addr;
+    size_t local_receiver_channel_buffer_address;
+
+    size_t termination_signal_ptr;
+
+    // Semaphore IDs
+    // this is the receiver channel's local sem for flow controlling with downstream fabric sender
+    std::optional<size_t> receiver_channel_downstream_flow_control_semaphore_id;
+    size_t sender_channel_0_flow_control_semaphore_id;
+    size_t sender_channel_1_flow_control_semaphore_id;
+    size_t sender_channel_0_connection_semaphore_id;
+    size_t sender_channel_1_connection_semaphore_id;
+    size_t sender_channel_0_buffer_index_semaphore_id;
+    size_t sender_channel_1_buffer_index_semaphore_id;
+    size_t receiver_channel_local_buffer_index_addr;
+
+    std::optional<size_t> downstream_edm_noc_x;
+    std::optional<size_t> downstream_edm_noc_y;
+    std::optional<size_t> downstream_edm_buffer_base_address;
+    std::optional<size_t> downstream_edm_semaphore_address;
+    std::optional<size_t> downstream_edm_worker_registration_address;
+    std::optional<size_t> downstream_edm_worker_location_info_address;
+    std::optional<size_t> downstream_sender_channel_buffer_index_semaphore_id;
+};
+
+
+struct edm_termination_info_t {
+    uint32_t distance;
+    uint32_t edm_noc_x;
+    uint32_t edm_noc_y;
+    uint32_t termination_addr;
+};
+
+struct EdmLineFabricOpInterface {
+    enum Direction {
+        // Ascending chips in the sequence
+        FORWARD,
+
+        // Descending chips in the sequence
+        BACKWARD,
+    };
+
+    // Device ID -> EDM Builders
+    std::unordered_map<size_t, std::vector<FabricEriscDatamoverBuilder>> edm_builders_forward_direction;
+    std::unordered_map<size_t, std::vector<FabricEriscDatamoverBuilder>> edm_builders_backward_direction;
+
+    // Device ID -> link index
+    std::unordered_map<size_t, size_t> next_forward_direction_edm_available;
+    std::unordered_map<size_t, size_t> next_backward_direction_edm_available;
+
+    std::vector<Device*> device_sequence;
+    std::vector<Program*> programs;
+
+    size_t num_links;
+
+    //   The constructor will assemble/connect the line across the specified device sequence, for all available links.
+    EdmLineFabricOpInterface (std::vector<Device*> const& device_sequence, std::vector<Program*> const& program_sequence, std::optional<size_t> desired_num_links = std::nullopt);
+
+
+    // Will create a connection adapter for a worker which can be used to pass args to the worker kernel talking to the
+    // corresponding fabric endpoint. This interface will guarantee unique connections only so requesting more unique connections
+    // than available will result in an error.
+    SenderWorkerAdapterSpec uniquely_connect_worker(Device* device, Direction direction);
```

Comment:
Can we remove "uniquely" from name? I understand that this is what it does, but I wonder how critical it is not expose it in the api name.

Reply:
for my sanity as a user of this, I would like to say no :) Realistically it will only be 2 - 3 of us using this API for the next 2-3 months before we deprecate